### ISSUE TITLE
Backfill `PartitionHashId` values in a Postgres migration

### ISSUE THREAD

**Issue by carols10cents** [influxdata/influxdb_iox#8525]
I thought it would be possible, @crepererum suggested it would be, and now I'm not so sure.

Here's the Rust code where we generate `PartitionHashId` values from a `TableId(i64)` and `PartitionKey(Arc<str>)`: https://github.com/influxdata/influxdb_iox/blob/38c3229e350b1f07815a37c10573f8c50e719543/data_types/src/partition.rs#L262-L283

In words: it's the SHA256 of the big-endian `TableId` bytes concatenated with the `PartitionKey`'s bytes.

Here's a sample to test with: https://github.com/influxdata/influxdb_iox/blob/38c3229e350b1f07815a37c10573f8c50e719543/data_types/src/partition.rs#L484-L492

For ease of copy pasting:
    - `TableId` 5
    - `PartitionKey` `"2023-06-08"`
    - generates the `PartitionHashId` (in hex) `ebd1041daa7c644c99967b817ae607bdcb754c663f2c415f270d6df720280f7a`.

[Postgres has a `sha256` function](https://www.postgresql.org/docs/current/functions-binarystring.html). Cool. It takes a `bytea`.

Getting a `bytea` of a `character varying` (the type of `partition_key` in Postgres) is easy: 

```
select '2023-06-08'::bytea;
         bytea
------------------------
 \x323032332d30362d3038
```

Getting a `bytea` of the big-endian representation of a `bigint` (the type of `table_id` in Postgres) is.... not easy. Here's what I've found:

- [Conversions through the hex representation??!?!](https://stackoverflow.com/questions/52327354/how-to-convert-bit-type-to-bytea-type-in-postgresql?rq=3) I've tried a few variations of this and haven't had luck
- [Multiple calls to `get_byte`?!?!?](https://www.postgresql.org/message-id/456F3E56-7596-4310-8FE5-BB06CEA561AA%40pgedit.com)
- ["The short answer is thus still that there is no sane consistent cast from int to bytea. If you want a bytea then it's really up to you to contruct the byte array."](https://www.postgresql.org/message-id/flat/474EEF16.5000705%40dunslane.net#36d6f9a791397247903e87fc0ee4558a) with bonus condescension "You don't know what't endianness is, do you?" and admonitions about top-posting

Here's what I thought should work but doesn't:

```
select decode(lpad(to_hex(5), 8, '0'), 'hex');
   decode
------------
 \x00000005

select sha256(decode(lpad(to_hex(5), 8, '0'), 'hex') || '2023-06-08'::bytea);
                               sha256
--------------------------------------------------------------------
 \xb58661cf24a256af9a1dfce168c6263d8e7a57625718bfe58ac4feba49f73d33
(1 row)
```

Help. Also I'm out the whole week of Aug 21. Just wanted to put this in for people to think about in the meantime.

**Comment by crepererum** [comment]
I'll have a look.

**Comment by crepererum** [comment]
@carols10cents was sooo close and I've tripped over this as well: hex encoding has 4 bits per characters (so half a byte), so `lpad` has to pad the hex string to 2*8=16 characters. With that change, it works:

```text
postgres=# select sha256(decode(lpad(to_hex(5), 16, '0'), 'hex') || '2023-06-08'::bytea);
                               sha256
--------------------------------------------------------------------
 \xebd1041daa7c644c99967b817ae607bdcb754c663f2c415f270d6df720280f7a
(1 row)
```

I've also searched the internet for other options but could find any that work w/o extensions. Another promising one is to cast to `bit(n)` (e.g. `(5::bigint)::bit(64)`), however I couldn't find a conversion from `bit(64)` to `bytea` :shrug:
